app/src/org/commcare/android/framework/CommCareActivity.java

```diff
@@ -141,6 +144,52 @@ protected void onCreate(Bundle savedInstanceState) {
         mGestureDetector = new GestureDetector(this, this);
     }
 
+    protected void requestMajorLayoutUpdates() {
+        final View decorView = getWindow().getDecorView();
+        decorView.getViewTreeObserver().addOnGlobalLayoutListener(new ViewTreeObserver.OnGlobalLayoutListener() {
+
+            int mPreviousDecorViewFrameHeight = 0;
+
+            @Override
+            public void onGlobalLayout() {

```

Comment:
If I'm understanding correctly: every time we setup the layout again, we get the currently available amount of space for the actual form. If this has changed by more than 100px we reason that the keyboard position has changed, and so update the space allotment for the group label. Is that the correct? If so, is there any reason not to check directly if the keyboard is up or not?